src/ProducerEventProcessor/ProducerEventProcessor.cpp

```diff
@@ -169,8 +172,56 @@ class ProducerEventProcessorImpl : public ProducerEventProcessor {
   // <producer_id, producer_string_id> -> client_string_id
   absl::flat_hash_map<std::pair<uint64_t, uint64_t>, uint64_t>
       producer_interned_string_id_to_client_string_id_;
+
+  // Needed to allow merging between call stacks and tracepoints, see design doc:
+  // http://go/stadia-orbit-tracepoint-callstack.
+  // NOTE: A thread state slice always gets constructed using two tracepoint events. It is always
+  // the begin tracepoint event that results in the ThreadStateSliceCallstack, so we will always
+  // see the ThreadStateSliceCallstack before we see the matching ThreadStateSlice. Thus, we do not
+  // need to save the thread state slices to be merged with a callstack later.
+  absl::flat_hash_map<std::pair<uint32_t, uint64_t>, uint64_t>
+      thread_state_slice_tid_and_begin_timestamp_to_callstack_id_;
 };
 
+void ProducerEventProcessorImpl::MergeThreadStateSliceWithCallstackAndTransferOwnership(
+    ThreadStateSlice* thread_state_slice) {
+  uint64_t begin_timestamp =
+      thread_state_slice->end_timestamp_ns() - thread_state_slice->duration_ns();
+  // Callstacks on thread state slices always origin from the tracepoint that corresponds to the
+  // slice's begin. Thus, if we see a thread state slice waiting for the callstack to be added,
+  // we know that we have already seen the corresponding callstack.
+  // Also, even if we were missing the end tracepoint, we are not leaking memory in our callstack
+  // map--the SwitchesStatesNamesVisitor takes care of that--and will eventually create a
+  // thread state slice for that begin tracepoint (worst case at the end of profiling).
+  auto thread_state_slice_callstack_it =
+      thread_state_slice_tid_and_begin_timestamp_to_callstack_id_.find(
+          {thread_state_slice->tid(), begin_timestamp});
+
+  // There are rare situations where we do not have a callstack even though the thread state slice
+  // is waiting for it. Mostly, because unwinding failed completely.
+  // Let's "fail" gracefully here, by not assigning a callstack.
+  if (thread_state_slice_callstack_it ==
+      thread_state_slice_tid_and_begin_timestamp_to_callstack_id_.end()) {
+    thread_state_slice->set_switch_out_or_wakeup_callstack_status(ThreadStateSlice::kNoCallstack);
```

Comment:
Since we had an `ORBIT_CHECK` here before, what do you think of a `ORBIT_ERROR` now, if we still don't expect this to happen? It would basically be the counterpart of `ORBIT_ERROR("Unwound callstack has no frames");` in `UprobesUnwindingVisitor`.

Reply:
I like it. Done.